src/main/java/org/openrewrite/java/migrate/util/OptionalStreamRecipe.java

```diff
@@ -0,0 +1,133 @@
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.migrate.util;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.tree.Comment;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JRightPadded;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Space;
+import org.openrewrite.java.tree.TextComment;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class OptionalStreamRecipe extends Recipe {
+  @Override
+  public String getDisplayName() {
+    return "Stream<Optional> idiom recipe";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Migrate Java 8 optional stream idiom .filter(Optional::isPresent).map(Optional::get) to Java 11 .flatMap(Optional::stream).";
+  }
+
+  @Override
+  public TreeVisitor<?, ExecutionContext> getVisitor() {
+    return new OptionalStreamVisitor();
+  }
+
+  private static class OptionalStreamVisitor extends JavaIsoVisitor<ExecutionContext> {
+    private static final MethodMatcher mapMatcher = new MethodMatcher("java.util.stream.Stream map(..)");
+    private static final MethodMatcher filterMatcher = new MethodMatcher("java.util.stream.Stream filter(..)");
+    private static final MethodMatcher optionalGetMatcher = new MethodMatcher("java.util.Optional get()");
+    private static final MethodMatcher optionalIsPresentMatcher = new MethodMatcher("java.util.Optional isPresent()");
+    private static final JavaTemplate template =
+        JavaTemplate.builder("#{any(java.util.stream.Stream)}.flatMap(Optional::stream)")
+            .imports("java.util.Optional")
+            .build();
+
+    @Override
+    public J.MethodInvocation visitMethodInvocation(J.MethodInvocation invocation, ExecutionContext ctx) {
+      return getInvocation(invocation, mapMatcher)
+          .flatMap(mapInvocation -> getInvocation(mapInvocation.getSelect(), filterMatcher)
+          .flatMap(filterInvocation -> getReference(mapInvocation.getArguments().get(0), optionalGetMatcher)
+          .flatMap(optionalGetReference -> getReference(filterInvocation.getArguments().get(0), optionalIsPresentMatcher)
+          .flatMap(optionalIsPresentReference -> {
```

Comment:
I introduced chain of flatMaps because a series of nested ifs was very hard to read. I have no problem with reverting this change, but have a look at this change: https://github.com/openrewrite/rewrite-migrate-java/pull/315/commits/8550af0825d893af9ce4cd3dddb60a051f0f112d#diff-eadd116ad853efe772ca5fb5dec23f6417a410e1c5d6aa59b3d498057e335dc8L49
and tell which version do you prefer.

Reply:
I wouldn't necessarily use nested if statements; we can likely use early returns to keep the structure left aligned and relatively easy to follow. Let me know if you'd like to pass that challenge back to me instead; not trying to make you jump through hoops to get this through; we appreciate the work you've put in already! :)

Reply:
Added this just now with early returns.